src/HDF5.jl

```diff
@@ -1269,66 +1284,121 @@ function read(parent::Union{HDF5File, HDF5Group}, name::String; pv...)
     val
 end
 
+function read(parent::Union{HDF5File, HDF5Group}, name_type_pair::Pair{String, DataType}; pv...)
+    obj = getindex(parent, name_type_pair[1]; pv...)
+    val = read(obj, name_type_pair[2])
+    close(obj)
+    val
+end
+
 # "Plain" (unformatted) reads. These work only for simple types: scalars, arrays, and strings
 # See also "Reading arrays using getindex" below
 # This infers the Julia type from the HDF5Datatype. Specific file formats should provide their own read(dset).
 const DatasetOrAttribute = Union{HDF5Dataset, HDF5Attribute}
+
 function read(obj::DatasetOrAttribute)
-    T = hdf5_to_julia(obj)
-    read(obj, T)
+  dtype = datatype(obj)
+  T = get_jl_type(dtype)
+  read(obj, T)
 end
-# Read scalars
-function read(obj::DatasetOrAttribute, ::Type{T}) where {T<:Union{HDF5Scalar, Complex{<:HDF5Scalar}}}
-    x = read(obj, Array{T})
-    x[1]
+
+function getindex(dset::HDF5Dataset, I...)
+  dtype = datatype(dset)
+  T = get_jl_type(dtype)
+  read(dset, T, I...)
 end
-# Read array of scalars
-function read(obj::DatasetOrAttribute, ::Type{Array{T}}) where {T<:Union{HDF5Scalar, Complex{<:HDF5Scalar}}}
-    if isnull(obj)
-        return T[]
+
+# could be replaced by filter(i -> !isa(i, Int), args) in julia 1.4
```

Comment:
`filter(i -> !isa(i, Int), args)` works  for me on Julia 1.3 which is the minimum version stated for HDF5.jl

Reply:
I think the filter method for tuples was only added in 1.4. This is what I get in 1.3.

```
filter(i -> !isa(i, Int), (1, 1:10, 1))
ERROR: MethodError: no method matching filter(::var"#3#4", ::Tuple{Int64,UnitRange{Int64},Int64})
Closest candidates are:
  filter(::Any, ::Array{T,N}) where {T, N} at array.jl:2342
  filter(::Any, ::BitArray) at bitarray.jl:1710
  filter(::Any, ::AbstractArray) at array.jl:2354
  ...
```

Reply:
Ah sorry for some reason I tested it on Array types, which is why I was surprised.

Reply:
Let's  just use https://github.com/JuliaLang/Compat.jl/releases/tag/v3.1.0 instead of this hack?